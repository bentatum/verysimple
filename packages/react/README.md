# @verysimple/react

This is the `@verysimple/react` package