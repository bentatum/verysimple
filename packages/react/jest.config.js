/** @type {import('jest').Config} */
module.exports = {
  testEnvironment: "jsdom",
  setupFilesAfterEnv: ["@testing-library/jest-dom/extend-expect"],
  collectCoverage: true,
  coverageThreshold: {
    global: {
      statements: 75,
      branches: 72,
      functions: 75,
      lines: 75,
    },
  },
  coverageReporters: ["text-summary", "json-summary"],
  collectCoverageFrom: ["src/**/*.{ts,tsx}", "!**/*.d.ts"],
  coverageDirectory: "coverage",
  moduleNameMapper: {
    "@/test/(.*)": "<rootDir>/test/$1",
    "@/(.*)$": "<rootDir>/src/$1",
  },
  modulePathIgnorePatterns: ["<rootDir>/dist/"],
  prettierPath: null,
};
