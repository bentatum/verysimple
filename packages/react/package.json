{
  "name": "@verysimple/react",
  "version": "8.1.1",
  "description": "A very simple component library",
  "main": "dist/index.js",
  "files": [
    "dist"
  ],
  "author": "Ben Tatum <mail@benjtatum.com>",
  "license": "MIT",
  "repository": "https://github.com/bentatum/verysimple",
  "scripts": {
    "test": "jest",
    "prebuild": "rm -rf dist",
    "build": "tsc --project tsconfig.build.json && tsc-alias"
  },
  "peerDependencies": {
    "react": ">=16",
    "react-dom": ">=16"
  },
  "devDependencies": {
    "@babel/core": "^7.20.12",
    "@babel/preset-env": "^7.19.1",
    "@babel/preset-react": "^7.18.6",
    "@babel/preset-typescript": "^7.18.6",
    "@testing-library/dom": "^8.20.0",
    "@testing-library/jest-dom": "^5.16.5",
    "@testing-library/react": "^13.4.0",
    "@testing-library/user-event": "^14.4.3",
    "@types/inquirer": "latest",
    "@types/jest": "^29.2.3",
    "@types/react": "^18.0.21",
    "@types/testing-library__jest-dom": "^5.14.5",
    "babel-jest": "^29.0.3",
    "babel-loader": "^8.2.5",
    "classnames": "^2.3.2",
    "framer-motion": "^10.15.0",
    "jest": "^29.0.3",
    "jest-environment-jsdom": "^29.0.3",
    "react": "^18.2.0",
    "react-dom": "^18.2.0",
    "tailwindcss": "^3.1.8",
    "tsc-alias": "^1.8.2",
    "typescript": "^4.9.3"
  },
  "volta": {
    "node": "20.10.0",
    "yarn": "1.22.21"
  },
  "dependencies": {
    "@headlessui/react": "^2.1.9",
    "@heroicons/react": "^2.0.13",
    "clsx": "^2.1.0",
    "framer-motion": "^10.15.0"
  }
}
