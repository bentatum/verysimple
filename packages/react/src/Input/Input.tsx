import classNames from "classnames";
import {
  ComponentPropsWithoutRef,
  ComponentPropsWithRef,
  ElementType,
  forwardRef,
  ReactNode,
} from "react";
import { fieldPaddingClassNames } from "@/helpers/fieldPaddingClassNames";
import { shadowTransitionClassNames } from "@/helpers/shadowTransitionClassNames";
import { InputSize } from "@/types";
import { fieldSizeClassNames } from "@/helpers/fieldSizeClassNames";
import { fieldBorderRadiusClassNames } from "..";

export interface InputContainerClassnameFunctionProps {
  hasFocus: boolean;
}

export interface InputContainerProps extends ComponentPropsWithoutRef<"div"> {}

export interface InputProps extends ComponentPropsWithRef<"input"> {
  adornmentLeft?: ReactNode;
  adornmentRight?: ReactNode;
  error?: boolean;
  fullWidth?: boolean;
  as?: ElementType;
  disabled?: boolean;
  inputSize?: InputSize;
  containerProps?: InputContainerProps;
  readOnly?: boolean;
}

export const Input = forwardRef<HTMLInputElement, InputProps>(
  (
    {
      adornmentLeft,
      adornmentRight,
      children,
      className = "",
      error,
      fullWidth,
      as = "input",
      inputSize = "md",
      readOnly,
      containerProps,
      disabled,
      ...props
    },
    ref
  ) => {
    const InputComponent = as;
    const containerClassname = containerProps?.className || "";
    return (
      <div
        {...containerProps}
        className={classNames(
          "inline-flex items-center overflow-hidden",
          fieldBorderRadiusClassNames(containerClassname),
          {
            "border my-border": !containerClassname.includes("border-"),
            "w-full": fullWidth,
            [shadowTransitionClassNames(className)]: !disabled,
            "shadow-sm focus-within:shadow-md hover:shadow-md disabled:shadow-none":
              !disabled,
          },
          containerClassname
        )}
      >
        <div className="w-full relative inline-flex items-center">
          <InputComponent
            ref={ref}
            disabled={disabled}
            className={classNames(
              fieldSizeClassNames(inputSize),
              fieldPaddingClassNames(inputSize, className),
              "focus:outline-none appearance-none",
              "placeholder-zinc-400",
              "disabled:bg-zinc-200 dark:disabled:bg-zinc-600 disabled:cursor-not-allowed",
              {
                "w-full": fullWidth,
                "cursor-default": readOnly,
                "my-bg-primary": !className.match(/bg-/),
                "text-left dark:text-white dark:disabled:text-zinc-300 text-black":
                  !className.match(/text-/),
              },
              className
            )}
            {...props}
          >
            {children}
          </InputComponent>
          {adornmentLeft}
          {adornmentRight}
        </div>
      </div>
    );
  }
);

Input.displayName = "Input";

export default Input;
