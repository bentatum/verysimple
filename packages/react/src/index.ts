export * from "./ButtonBase";
export * from "./Card";
export * from "./Dialog";
export * from "./FilledButtonBase";
export * from "./FilledIconButton";
export * from "./GradientBorder";
export * from "./helpers";
export * from "./IconButton";
export * from "./NormalIconButton";
export * from "./OutlinedButtonBase";
export * from "./OutlinedIconButton";
export * from "./Spinner";
export * from "./types";
