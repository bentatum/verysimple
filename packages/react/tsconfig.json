{
  "extends": "@verysimple/tsconfig/react-library.json",
  "compilerOptions": {
    "outDir": "./dist",
    "paths": {
      "@/*": ["./src/*"]
    }
  },
  "include": ["src/**/*.ts", "src/**/*.tsx"],
  "exclude": ["node_modules"]
}
