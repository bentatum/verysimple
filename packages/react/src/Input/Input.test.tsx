import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { createRef } from "react";
import Input from "./Input";

beforeEach(() => {
  jest.spyOn(console, "error").mockImplementation(() => {});
});

describe("Input", () => {
  it("should render", () => {
    const { container } = render(<Input />);
    expect(container).toMatchSnapshot();
  });
  describe("full width", () => {
    it("sets the container to w-full", () => {
      const { container } = render(<Input className="w-full" />);
      expect(container.firstChild).toHaveClass("w-full");
    });
    it("adds the w-full class to the textbox", () => {
      render(<Input className="w-full" />);
      expect(screen.getByRole("textbox")).toHaveClass("w-full");
    });
  });

  it("should not have shadow classnames when disabled", () => {
    const { container } = render(<Input disabled />);
    expect(container.firstChild).not.toHaveClass(
      "shadow-sm",
      "hover:shadow-md"
    );
    expect(container).toMatchSnapshot();
  });

  it("should call focus and blur handlers", async () => {
    const user = userEvent.setup();
    const onFocus = jest.fn();
    const onBlur = jest.fn();
    render(<Input onFocus={onFocus} onBlur={onBlur} />);
    await user.click(screen.getByRole("textbox"));
    expect(onFocus).toHaveBeenCalledTimes(1);
    await user.click(document.body);
    expect(onBlur).toHaveBeenCalledTimes(1);
  });

  it("should not have a gradient border if disable focus style is true", async () => {
    const user = userEvent.setup();
    const { container } = render(<Input disableFocusStyle />);
    await user.click(screen.getByRole("textbox"));
    expect(container.firstChild).not.toHaveStyle({
      backgroundClip: "content-box, border-box",
      background:
        "linear-gradient(var(--my-bg-primary), var(--my-bg-primary)) 0 0 / 100% 100% no-repeat, linear-gradient(to right, var(--my-gradient-start) 0%, var(--my-gradient-stop) 100%)",
      padding: "1px",
    });
    expect(container).toMatchSnapshot();
    await user.click(document.body);
    expect(container).toMatchSnapshot();
  });

  it("forwards the ref", () => {
    const ref = createRef<HTMLInputElement>();
    render(<Input ref={ref} />);
    expect(ref.current).toBeInTheDocument();
  });

  it("has a display name", () => {
    expect(Input.displayName).toBe("Input");
  });

  test("container props classname as a string", () => {
    const { container } = render(
      <Input containerProps={{ className: "test" }} />
    );
    expect(container.firstChild).toHaveClass("test");
  });
});
