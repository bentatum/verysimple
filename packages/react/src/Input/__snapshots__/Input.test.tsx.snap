// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input should not have a gradient border if disable focus style is true 1`] = `
<div>
  <div
    class="rounded inline-flex items-center overflow-hidden border my-border transition-shadow ease-in duration-100 shadow-sm focus-within:shadow-md hover:shadow-md disabled:shadow-none"
  >
    <div
      class="w-full relative inline-flex items-center"
    >
      <input
        class="text-base min-h-[44px] px-4 focus:outline-none appearance-none placeholder-gray-400 disabled:bg-gray-200 dark:disabled:bg-gray-600 disabled:cursor-not-allowed bg-background text-left dark:text-white dark:disabled:text-gray-300 text-black"
      />
    </div>
  </div>
</div>
`;

exports[`Input should not have a gradient border if disable focus style is true 2`] = `
<div>
  <div
    class="rounded inline-flex items-center overflow-hidden border my-border transition-shadow ease-in duration-100 shadow-sm focus-within:shadow-md hover:shadow-md disabled:shadow-none"
  >
    <div
      class="w-full relative inline-flex items-center"
    >
      <input
        class="text-base min-h-[44px] px-4 focus:outline-none appearance-none placeholder-gray-400 disabled:bg-gray-200 dark:disabled:bg-gray-600 disabled:cursor-not-allowed bg-background text-left dark:text-white dark:disabled:text-gray-300 text-black"
      />
    </div>
  </div>
</div>
`;

exports[`Input should not have shadow classnames when disabled 1`] = `
<div>
  <div
    class="rounded inline-flex items-center overflow-hidden border my-border"
  >
    <div
      class="w-full relative inline-flex items-center"
    >
      <input
        class="text-base min-h-[44px] px-4 focus:outline-none appearance-none placeholder-gray-400 disabled:bg-gray-200 dark:disabled:bg-gray-600 disabled:cursor-not-allowed bg-background text-left dark:text-white dark:disabled:text-gray-300 text-black"
        disabled=""
      />
    </div>
  </div>
</div>
`;

exports[`Input should render 1`] = `
<div>
  <div
    class="rounded inline-flex items-center overflow-hidden border my-border transition-shadow ease-in duration-100 shadow-sm focus-within:shadow-md hover:shadow-md disabled:shadow-none"
  >
    <div
      class="w-full relative inline-flex items-center"
    >
      <input
        class="text-base min-h-[44px] px-4 focus:outline-none appearance-none placeholder-gray-400 disabled:bg-gray-200 dark:disabled:bg-gray-600 disabled:cursor-not-allowed bg-background text-left dark:text-white dark:disabled:text-gray-300 text-black"
      />
    </div>
  </div>
</div>
`;
